src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs

```diff
@@ -192,39 +198,38 @@ private void CommonInit()
                 }
                 else
                 {
-                    SearchResult searchResult = CreateSearchResult(_searchData, data);
-                    if (_resultHandler.IsResultIncluded(searchResult))
+                    TSource result;
+                    if (IsResultIncluded(ref data, out result))
                     {
-                        current = _resultHandler.CreateObject(searchResult);
+                        current = result;
                     }
                 }
             }
-            // for AllDirectories, we first recurse into dirs, so cleanup and add searchData 
+            // for AllDirectories, we first recurse into dirs, so cleanup and add searchData
             // to the stack
             else
             {
                 _hnd.Dispose();
+                _searchStack = new List<SearchData>();
```

Comment:
I noticed this as well. It's using `List<T>` purely for historical reasons since this code was originally in `mscorlib.dll` and `Stack<T>` was in `System.dll`.

The original behavior (pre .NET 4) ended up producing a very strange ordering that was very different from the default order of `dir /s`. When we added the `Enumerate*` methods in .NET 4, we deliberately changed the previous behavior to match more closely with the default order of `dir /s` (although, the order is not guaranteed). I guess it wasn't implemented as efficiently has it could have been. We were able to make this change in .NET 4 without it being a breaking change because the order isn't guaranteed and its stated so in the docs. We should be able to change the order again, but my preference would be to maintain the current behavior as its a nicer ordering.

I went ahead and updated the code to operate at the end of the list, to avoid all the shifting. For adding the subdirectories, I changed it the way you suggested, adding each directory at the end and then calling Reverse on just that portion that was added. Note: I found that `List<PathPair>.Reverse` was slow (it looks like the underlying implementation in `Array.Reverse` ends up using a code path in this case that boxes each swapped element), so I implemented a custom non-boxing Reverse. The custom Reverse is ~28x faster than `List<PathPair>.Reverse`.